mbq/client/contrib/permissions.py

```diff
@@ -0,0 +1,216 @@
+import json
+import logging
+import urllib
+from copy import copy
+from typing import Dict, List, Optional, Union, cast, overload
+from uuid import UUID
+
+from typing_extensions import Literal, Protocol
+
+from .. import ServiceClient
+
+logger = logging.getLogger(__name__)
+
+UUIDType = Union[str, UUID]
+# External type returned from internal and external OS Core clients. Keys
+# are the org refs if UUIDs, or {ref_type}:{ref_id} if legacy int types.
+# Values are lists of scopes. Should also include a "global" literal key.
+FetchedPermissionsDoc = Dict[str, List[str]]
+# Internal type stored in the cache. Keys are cache keys with prefixes, colons, etc.
+# Values are pipe-delimited strings with an additional pipe on the end.
+CachedPermissionsDoc = Dict[str, str]
+RefType = Union[Literal["company", "vendor"]]
+
+
+class OSCoreClient(Protocol):
+    def fetch_permissions_for_org(
+        self, person_id: str, org_ref: UUIDType
+    ) -> FetchedPermissionsDoc:
+        ...
+
+    def fetch_permissions_for_location(
+        self, person_id: str, location_id: int, location_type: RefType
+    ) -> FetchedPermissionsDoc:
+        ...
+
+
+class OSCoreServiceClient:
+    def __init__(self, client: ServiceClient):
+        # The copying and munging here is attempting to deal with differences
+        # in how the individual ServiceClients are configured. We can get rid
+        # of it if we standardize within the services.
+
+        # Note that we are doing a shallow copy so the Authenticator instance
+        # will be shared
+        self.client = copy(client)
+
+        def _post_process_response(data):
+            if isinstance(data, dict):
+                objects = data.get("objects", data)
+                objects["global"] = []
+                return objects
+            else:
+                return data
+
+        self.client._post_process_response = _post_process_response
+        self.client._headers = None
+
+        parsed = urllib.parse.urlparse(self.client._api_url)
+        self.client._api_url = f"{parsed.scheme}://{parsed.netloc}"
+
+    def fetch_permissions_for_org(
```

Comment:
Could we actually possibly just name this `fetch_permissions`? Technically the `org_ref` could be either an `org` or a `location`. I know the naming sucks, but I think just leaving it out would be better than having the explicit `_for_org`, What do you think?

Reply:
Yeah you're right